src/index.js

```diff
@@ -156,13 +156,23 @@ class IOpipeWrapperClass {
         this.context,
         this.callback
       );
-      if (result && result.then && result.catch) {
-        return new Promise(() => this.callback(null, () => result))
-          .then(value => value)
-          .catch(err => {
-            this.sendReport(err, () => this.originalCallback(err));
-            return err;
-          });
+      if (
+        result &&
+        typeof result.then === 'function' &&
+        typeof result.catch === 'function'
+      ) {
+        return new Promise((resolve, reject) => {
+          return result
+            .then(value => {
+              this.context.succeed(value);
+              return this.callback(null, () => resolve(value));
+            })
+            .catch(err => {
+              this.context.fail(err);
+              this.sendReport(err, () => this.originalCallback(err));
```

Comment:
Is this the correct order? Doesn't `context.fail()` hand back to AWS?

Reply:
Ultimately, yes--context.fail's our wrapper around originalContext.fail, where it should hand back to AWS.  Our wrapper *also* invokes sendReport on .fail, but doesn't invoke originalCallback with an error.  

In testing, I haven't seen differences in behavior when the order changes, nor any difference from sendReport invoking originalContext.fail instead of this sendReport invoking originalCallback with an error...so I'm in favor of removing the redundancy and relying on context.fail here.

Reply:
Could we consolidate this into the `succeed()` and `fail()` methods? Send the report and call the oroginal context there?

Reply:
I'd think so.  Testing that approach...

Reply:
This approach can work, though if we do that, then we either have to abandon using resolve/reject (which would complete the promise before the callback and sendReport), or pass the resolve/reject to the succeed/fail methods, which starts adding responsibilities to these functions.  

It seems better to leverage the sendReport in context.fail to reduce duplication, and leave the promise/resolve pattern in place here, unless we want to abandon the resolve/reject...though currently, the result needs to be wrapped in a promise in order to execute in the next hook, and if we're using a promise, we should at least resolve it. (Reject completes the promise before the error can be reported.)